File: TMS.API/Models/Component.cs
using System;
using System.Collections.Generic;

namespace TMS.API.Models
{
    public partial class Component
    {
        public Component()
        {
            EntityRef = new HashSet<EntityRef>();
        }

        public string Id { get; set; }
        public string FieldName { get; set; }
        public string Order { get; set; }
        public string ComponentType { get; set; }
        public string ComponentGroupId { get; set; }
        public string DataSourceFilter { get; set; }
        public string ReferenceId { get; set; }
        public string FormatData { get; set; }
        public string FormatEntity { get; set; }
        public string PlainText { get; set; }
        public string Column { get; set; }
        public string Offset { get; set; }
        public string Row { get; set; }
        public bool CanSearch { get; set; }
        public bool CanCache { get; set; }
        public string Precision { get; set; }
        public string GroupBy { get; set; }
        public string GroupFormat { get; set; }
        public string Label { get; set; }
        public bool ShowLabel { get; set; }
        public string Icon { get; set; }
        public string ClassName { get; set; }
        public string Style { get; set; }
        public string ChildStyle { get; set; }
        public string HotKey { get; set; }
        public string RefClass { get; set; }
        public string Events { get; set; }
        public bool Disabled { get; set; }
        public bool Visibility { get; set; }
        public bool Hidden { get; set; }
        public string Validation { get; set; }
        public bool Focus { get; set; }
        public string Width { get; set; }
        public string PopulateField { get; set; }
        public string CascadeField { get; set; }
        public string GroupEvent { get; set; }
        public string XsCol { get; set; }
        public string SmCol { get; set; }
        public string LgCol { get; set; }
        public string XlCol { get; set; }
        public string XxlCol { get; set; }
        public string DefaultVal { get; set; }
        public string DateTimeField { get; set; }
        public bool Active { get; set; }
        public DateTimeOffset InsertedDate { get; set; }
        public string InsertedBy { get; set; }
        public DateTimeOffset? UpdatedDate { get; set; }
        public string UpdatedBy { get; set; }
        public string RoleId { get; set; }
        public bool IgnoreSync { get; set; }
        public bool CanAdd { get; set; }
        public bool ShowAudit { get; set; }
        public bool IsPrivate { get; set; }
        public string IdField { get; set; }
        public string DescFieldName { get; set; }
        public string DescValue { get; set; }
        public string MonthCount { get; set; }
        public bool? IsDoubleLine { get; set; }
        public string Query { get; set; }
        public bool IsRealtime { get; set; }
        public string RefName { get; set; }
        public bool TopEmpty { get; set; }
        public bool IsCollapsible { get; set; }
        public string Template { get; set; }
        public string System { get; set; }
        public string PreQuery { get; set; }
        public string DisabledExp { get; set; }
        public bool FocusSearch { get; set; }
        public bool IsSumary { get; set; }
        public string FormatSumaryField { get; set; }
        public string OrderBySumary { get; set; }
        public bool ShowHotKey { get; set; }
        public string DefaultAddStart { get; set; }
        public string DefaultAddEnd { get; set; }
        public bool UpperCase { get; set; }
        public bool VirtualScroll { get; set; }
        public string Migration { get; set; }
        public string ListClass { get; set; }
        public string ExcelFieldName { get; set; }
        public bool LiteGrid { get; set; }
        public bool ShowDatetimeField { get; set; }
        public bool ShowNull { get; set; }
        public bool AddDate { get; set; }
        public bool FilterEq { get; set; }
        public string HeaderHeight { get; set; }
        public string BodyItemHeight { get; set; }
        public string FooterHeight { get; set; }
        public string ScrollHeight { get; set; }
        public string ScriptValidation { get; set; }
        public bool FilterLocal { get; set; }
        public bool HideGrid { get; set; }
        public string GroupReferenceId { get; set; }
        public string GroupReferenceName { get; set; }
        public string JoinTable { get; set; }
        public string GroupName { get; set; }
        public string ShortDesc { get; set; }
        public string Description { get; set; }
        public string FeatureId { get; set; }
        public string EntityId { get; set; }
        public string ComponentId { get; set; }
        public string TextAlign { get; set; }
        public bool HasFilter { get; set; }
        public bool Frozen { get; set; }
        public string FilterTemplate { get; set; }
        public bool Editable { get; set; }
        public string FormatExcell { get; set; }
        public string DatabaseName { get; set; }
        public string Summary { get; set; }
        public string SummaryColSpan { get; set; }
        public bool BasicSearch { get; set; }
        public bool IsExport { get; set; }
        public string OrderExport { get; set; }
        public string ShowExp { get; set; }
        public string MinWidth { get; set; }
        public string MaxWidth { get; set; }
        public bool AdvancedSearch { get; set; }
        public bool AutoFit { get; set; }
        public bool DisplayNone { get; set; }

        public virtual ComponentGroup ComponentGroup { get; set; }
        public virtual Entity Reference { get; set; }
        public virtual ICollection<EntityRef> EntityRef { get; set; }
    }
}

File: frontend/lib/models/component.js
import { Action } from './action';

/**
 * @typedef {import('../index').EditableComponent} EditableComponent
 * Represents a component.
 * @class
 */
export class Component {
    /** @type {string} */
    id = null;
    /** @type {string} */
    fieldName = null;
    /** @type {string} */
    populateFieldName = null;
    /** @type {number} */
    order = 0;
    /** @type {string | (...args) => EditableComponent} */
    componentType = null;
    /** @type {string} */
    componentGroupId = null;
    /** @type {string} */
    formatData = null;
    /** @type {string} */
    plainText = null;
    /** @type {number} */
    column = 0;
    /** @type {number} */
    offset = 0;
    /** @type {number} */
    row = 0;
    /** @type {boolean} */
    canSearch = false;
    /** @type {boolean} */
    canCache = false;
    /** @type {number} */
    precision = 0;
    /** @type {string} */
    groupBy = null;
    /** @type {string} */
    groupFormat = null;
    /** @type {string} */
    label = null;
    /** @type {boolean} */
    showLabel = false;
    /** @type {string} */
    icon = null;
    /** @type {string} */
    className = null;
    /** @type {string} */
    style = null;
    /** @type {string} */
    childStyle = null;
    /** @type {string} */
    hotKey = null;
    /** @type {string} */
    refClass = null;
    /** @type {string} */
    events = null;
    /** @type {boolean} */
    disabled = false;
    /** @type {boolean} */
    visibility = false;
    /** @type {string} */
    validation = null;
    /** @type {boolean} */
    focus = false;
    /** @type {string} */
    width = null;
    /** @type {string} */
    populateField = null;
    /** @type {string} */
    groupEvent = null;
    /** @type {number} */
    xsCol = 0;
    /** @type {number} */
    smCol = 0;
    /** @type {number} */
    lgCol = 0;
    /** @type {number} */
    xlCol = 0;
    /** @type {number} */
    xxlCol = 0;
    /** @type {string} */
    defaultVal = null;
    /** @type {string} */
    dateTimeField = null;
    /** @type {boolean} */
    active = false;
    /** @type {Date} */
    insertedDate = new Date();
    /** @type {string} */
    insertedBy = null;
    /** @type {Date} */
    updatedDate = new Date();
    /** @type {string} */
    updatedBy = null;
    /** @type {boolean} */
    canAdd = false;
    /** @type {boolean} */
    isPrivate = false;
    /** @type {number} */
    monthCount = 0;
    /** @type {string} */
    query = null;
    /** @type {boolean} */
    isRealtime = false;
    /** @type {string} */
    refName = null;
    /** @type {boolean} */
    topEmpty = false;
    /** @type {boolean} */
    isCollapsible = false;
    /** @type {string} */
    template = null;
    /** @type {string} */
    preQuery = null;
    /** @type {string} */
    disabledExp = null;
    /** @type {boolean} */
    focusSearch = false;
    /** @type {boolean} */
    isSumary = false;
    /** @type {string} */
    formatSumaryField = null;
    /** @type {string} */
    orderBySumary = null;
    /** @type {boolean} */
    showHotKey = false;
    /** @type {number} */
    defaultAddStart = 0;
    /** @type {number} */
    defaultAddEnd = 0;
    /** @type {boolean} */
    upperCase = false;
    /** @type {boolean} */
    virtualScroll = false;
    /** @type {string} */
    migration = null;
    /** @type {string} */
    listClass = null;
    /** @type {string} */
    excelFieldName = null;
    /** @type {boolean} */
    liteGrid = false;
    /** @type {boolean} */
    showDatetimeField = false;
    /** @type {boolean} */
    showNull = false;
    /** @type {boolean} */
    addDate = false;
    /** @type {boolean} */
    displayBadge = false;
    /** @type {boolean} */
    filterEq = false;
    /** @type {number} */
    headerHeight = 0;
    /** @type {number} */
    bodyItemHeight = 0;
    /** @type {number} */
    footerHeight = 0;
    /** @type {number} */
    scrollHeight = 0;
    /** @type {string} */
    scriptValidation = null;
    /** @type {boolean} */
    filterLocal = false;
    /** @type {boolean} */
    hideGrid = false;
    /** @type {string} */
    groupReferenceId = null;
    /** @type {string} */
    groupReferenceName = null;
    /** @type {string} */
    groupName = null;
    /** @type {string} */
    shortDesc = null;
    /** @type {string} */
    description = null;
    /** @type {string} */
    featureId = null;
    /** @type {string} */
    componentId = null;
    /** @type {string} */
    textAlign = null;
    /** @type {boolean} */
    hasFilter = false;
    /** @type {boolean} */
    frozen = false;
    /** @type {boolean} */
    frozenRight = false;
    /** @type {string} */
    filterTemplate = null;
    /** @type {boolean} */
    editable = false;
    /** @type {string} */
    formatExcell = null;
    /** @type {string} */
    databaseName = null;
    /** @type {string} */
    summary = null;
    /** @type {number} */
    summaryColSpan = 0;
    /** @type {boolean} */
    basicSearch = false;
    /** @type {string} */
    showExp = null;
    /** @type {string} */
    minWidth = null;
    /** @type {string} */
    maxWidth = null;
    /** @type {string} */
    tenantCode = null;
    /** @type {string} */
    orderBy = null;
    /** @type {string} */
    parentId = null;
    /** @type {string} */
    lang = null;
    /** @type {string} */
    name = null;
    /** @type {boolean} */
    isTab = false;
    /** @type {string} */
    tabGroup = null;
    /** @type {boolean} */
    isVertialTab = false;
    /** @type {boolean} */
    responsive = false;
    /** @type {number} */
    outerColumn = 0;
    /** @type {number} */
    xsOuterColumn = 0;
    /** @type {number} */
    smOuterColumn = 0;
    /** @type {number} */
    lgOuterColumn = 0;
    /** @type {number} */
    xlOuterColumn = 0;
    /** @type {number} */
    xxlOuterColumn = 0;
    /** @type {number} */
    badgeMonth = 0;
    /** @type {boolean} */
    isDropDown = false;
    /** @type {boolean} */
    isMultiple = false;
    /** @type {string} */
    html = null;
    /** @type {string} */
    css = null;
    /** @type {string} */
    javascript = null;
    /** @type {HTMLElement} */
    parentElement = null;
    /** @type {Component[]} */
    columns = [];
    /** @type {Component[]|string} */
    localQuery;
    /** @type {Component[]} */
    components;
    /** @type {any} */
    layout;
    /** @type {Function} */
    onClick;
    /** @type {string} */
    addRowExp;
    /** @type {string} */
    entityName;
    /** @type {string} */
    tableName;
    /** @type {string} */
    rowSpan;
    /** @type {string} */
    componentDefaultValueId;
    /** @type {string} */
    groupTypeId;
    canRead = false;
    canReadAll = false;
    canWrite = false;
    canWriteAll = false;
    canDelete = false;
    canDeleteAll = false;
    canCopy = false;
    canCopyAll = false;
    canDeactivate = false;
    canDeactivateAll = false;
    canExport = false;
    isPublic = false;
}